adsplanetnamepipe/utils/knowledge_graph.py

```diff
@@ -0,0 +1,133 @@
+from typing import List
+from collections import defaultdict
+
+from adsputils import setup_logging, load_config
+
+logger = setup_logging('utils')
+config = {}
+config.update(load_config())
+
+import networkx as nx
+
+from adsplanetnamepipe.utils.common import EntityArgs
+
+
+class KnowledgeGraph():
+
+    def __init__(self, args: EntityArgs, keywords: List[List['str']], special_keywords: List[List['str']]):
+        """
+
+        :param args:
+        :param keywords:
+        :param special_keywords:
+        """
+        self.args = args
+        self.score_format = '%.{}f'.format(config['PLANETARYNAMES_PIPELINE_FORMAT_SIGNIFICANT_DIGITS'])
+        self.graph = nx.Graph()
+
+        self.build_graph(keywords, special_keywords)
+
+    def build_graph(self, keywords_list, special_keywords_list):
+        """
+
+        :param keywords_list:
+        :param special_keywords_list: these have a higher weights
+        :return:
+        """
+        self.keyword_counts, self.pair_counts = self.count_keywords(keywords_list)
+        self.special_keyword_counts, self.special_pair_counts = self.count_keywords(special_keywords_list)
+
+        if len(self.keyword_counts) > 0:
+            self.create_graph()
+        else:
+            self.graph = None
+
+    def create_graph(self):
+        """
+
+        :return:
+        """
+        for (keyword_counts, pair_counts, weight) in zip([self.keyword_counts, self.special_keyword_counts],
+                                                         [self.pair_counts, self.special_pair_counts],
+                                                         [1, 2]):
+            keyword_list = sorted(keyword_counts.items(), key=lambda x: x[1], reverse=True)
+            pair_list = sorted(pair_counts.items(), key=lambda x: x[1], reverse=True)
+            # weights for regular keywords are 1, for special keywords are 2
+            self.add_nodes_and_edges(keyword_list, pair_list, weight)
+
+    def count_keywords(self, keywords_list):
+        """
+
+        :param keywords_list:
+        :return:
+        """
+        keyword_counts = defaultdict(int)
+        pair_counts = defaultdict(int)
+
+        for keywords in keywords_list:
+            for keyword in keywords:
```

Comment:
This will end up double counting pairs - not sure from the context if that's expected or not

Reply:
No, should not. I will check it out. Note that there are two lists
https://github.com/golnazads/ADSPlanetaryNamesPipeline/blob/main/adsplanetnamepipe/tests/unittests/stubdata/keywords.py#L2
one keywords from spacy/yake/wikidata and one phrases from nasa concept, eventually.

Reply:
checked this again, it does not double count in error. If you mean it counts both 'entropy' and 'diagnosis', also 'diagnosis' and 'entropy', that is by design since we want to count co occurrence in both direction.